src/rvsdg/egglog_optimizer/conditional_invariant_code_motion.egg

```diff
@@ -0,0 +1,50 @@
+; If (= x (prefix-containing-operand (Gamma ...) e)), then the first x outputs
+; of Gamma contain e.
+(function prefix-containing-Operand (Body Operand) i64 :merge (max old new))
+(rule
+    ((= gamma (Gamma pred inputs outputs))
+     (Body-contains-Operand gamma 0 e))
+    ((set (prefix-containing-Operand gamma e) 0))
+    :ruleset fast-analyses)
+(rule
+    ((= gamma (Gamma pred inputs outputs))
+     (= i (prefix-containing-Operand gamma e))
+     (Body-contains-Operand gamma i e))
+    ((set (prefix-containing-Operand gamma e) (+ i 1)))
+    :ruleset fast-analyses)
+
+; Limit blowup by only repeatedly hoisting so many times and by limiting the
+; number of branches we hoist from or max number of inputs we can add.
+(let max-hoists 1)
+(function hoist-count (Body) i64 :merge (max old new))
+(rule ((= gamma (Gamma pred inputs outputs)))
+      ((set (hoist-count gamma) 0))
+      :ruleset fast-analyses)
+(let max-branches 2)
```

Comment:
What's the point of max-branches and max-inputs? Doesn't max-hoists take care of the blowup?

Reply:
To limit the number of Gammas we try to hoist in the first place, sadly this these are also necessary

Reply:
I think the tradeoff is --
lower limits -> less performance interference with other's optimizations
higher limits -> people can more easily write rules relying on CICM

Because our performance is very bad now, I chose the former

Reply:
Hmm that's unfortunate. Okay
@clyben is experimenting with an analysis that finds nodes on the boundary between invariant and not invariant, and uses those. But okay for this PR